<font size="+3"><b>Assignment 2: Linear Models and Validation Metrics</b></font>

***
* **Full Name** = Alessandro Baldassarre
* **UCID** = 30163507
***

<font color='Blue'>In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.</font>

You can use the Table of Content on the left side of this notebook to efficiently navigate within this documents.

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|           **Part 2: Regression**           |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    0.5    |
| Step 3: Implement Machine Learning   Model |     1     |
|            Step 4: Validate Mode           |     1     |
|          Step 5: Visualize Results         |     1     |
|                  Questions                 |     2     |
|             Process Description            |     4     |
|  **Part 3:   Observations/Interpretation** |   **3**   |
|           **Part 4: Reflection**           |   **2**   |
|                  **Total**                 |   **30**  |
|                                            |           |
|                  **Bonus**                 |           |
|         **Part 5: Bonus Question**         |   **4**   |

# **Part 1: Classification (14.5 marks total)**

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|                  **Total**                 |  **14.5** |

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:


## **Step 0:** Import Libraries

In [1]:
import numpy as np
import pandas as pd

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [3]:
# TO DO: Import spam dataset from yellowbrick library
%pip install yellowbrick
from yellowbrick.datasets.loaders import load_spam
X, y = load_spam(return_dataset=False)

# TO DO: Print size and type of X and y
print(f'Size: {X.size}, Type: {type(X)}')
print(f'Size: {y.size}, Type: {type(y)}')

ModuleNotFoundError: No module named 'yellowbrick'

## **Step 2:** Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [ ]:
# TO DO: Check if there are any missing values and fill them in if necessary

# Check for missing values in dataset
missing_values_X = X.isnull().sum()

# Fill in missing values using mean of each column in dataset
X_filled = X.fillna(X.mean(), inplace=True)

For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [ ]:
# TO DO: Create X_small and y_small

from sklearn.model_selection import train_test_split

# Use train_test_split to allocate 5% of the data to X_small and y_small
X_small, X_unused, y_small, y_unused = train_test_split(X, y, train_size=0.05, random_state=42)

# Printing size of new x and y to confirm 5% of dataset is used
print("X_small size:", X_small.size)
print("y_small size:", y_small.size)

X_small size: 13110
y_small size: 230


## **Step 3:** Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets:
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

## **Step 4:** Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

## **Step 5:** Visualize Results (4 marks for steps 3-5)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [ ]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

lrm = LogisticRegression(max_iter=2000) # instantiating Logistic Regression Model

# 1. Implementing the lrm with the full dataset (X and y)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
lrm.fit(X_train, y_train)

# Calculating training accuracy
train_predictions = lrm.predict(X_train)
train_accuracy_full = accuracy_score(y_train, train_predictions)

# Calculating validation accuracy
validation_predictions = lrm.predict(X_test)
validation_accuracy_full = accuracy_score(y_test, validation_predictions)

# 2. Implementing the lrm using only the first two columns of X and y
X_2_cols = X.iloc[:, :2]
X_train_2, X_test_2, y_train_2, y_test_2 = train_test_split(X_2_cols, y, test_size=0.2, random_state=0)
lrm.fit(X_train_2, y_train_2)

# Calculating training accuracy
train_predictions_2 = lrm.predict(X_train_2)
train_accuracy_2_cols = accuracy_score(y_train_2, train_predictions_2)

# Calculating validation accuracy
validation_predictions_2 = lrm.predict(X_test_2)
validation_accuracy_2_cols = accuracy_score(y_test_2, validation_predictions_2)

# 3. Implementing the lrm with the smaller dataset (X_small and y_small)
X_train_small, X_test_small, y_train_small, y_test_small = train_test_split(X_small, y_small, test_size=0.2, random_state=0)
lrm.fit(X_train_small, y_train_small)

# Calculating training accuracy
train_predictions_small = lrm.predict(X_train_small)
train_accuracy_small = accuracy_score(y_train_small, train_predictions_small)

# Calculating validation accuracy
validation_predictions_small = lrm.predict(X_test_small)
validation_accuracy_small = accuracy_score(y_test_small, validation_predictions_small)

# Printing all accuracies for all datasets
print("Full dataset - Training accuracy:", train_accuracy_full)
print("Full dataset - Validation accuracy:", validation_accuracy_full)
print("First two columns - Training accuracy:", train_accuracy_2_cols)
print("First two columns - Validation accuracy:", validation_accuracy_2_cols)
print("Small dataset - Training accuracy:", train_accuracy_small)
print("Small dataset - Validation accuracy:", validation_accuracy_small)

Full dataset - Training accuracy: 0.9269021739130435
Full dataset - Validation accuracy: 0.9369565217391305
First two columns - Training accuracy: 0.6149456521739131
First two columns - Validation accuracy: 0.5934782608695652
Small dataset - Training accuracy: 0.9619565217391305
Small dataset - Validation accuracy: 0.8913043478260869


In [ ]:
# Prepare the data for the DataFrame
data = {
    "Data size": [
        X_train.size,  # Full dataset
        X_train_2.size,  # First two columns
        X_train_small.size  # Small dataset
    ],
    "Training accuracy": [
        train_accuracy_full,
        train_accuracy_2_cols,
        train_accuracy_small
    ],
    "Validation accuracy": [
        validation_accuracy_full,
        validation_accuracy_2_cols,
        validation_accuracy_small
    ]
}

# Create the DataFrame
results = pd.DataFrame(data, index=["Full dataset", "First two columns", "Small dataset"])

# Print the results DataFrame
results

Data size  Training accuracy  Validation accuracy
Full dataset          209760           0.926902             0.936957
First two columns       7360           0.614946             0.593478
Small dataset          10488           0.961957             0.891304

## **Questions (4 marks)**
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

<font color='Green'><b>YOUR ANSWERS HERE</b></font>
1. Using the full dataset for training would provide the most understanding to the model and as a result, higher accuracies are expected. In this case the full dataset had accuracies of 0.934 and 0.909 for training and validation respectively. Really high values compared to the first two columns. Only using the first two columns reduces the feature space so it's a 50/50. The columns could either contain really significant features or miss those really significant features. The training and validation accuracies dropped significantly compared to the full dataset with values at 0.620 and 0.589 respectively. As for using a smaller version of the dataset, it would allow the model to more easily fit to that dataset usually resulting in greater values in training accuracy (0.962 in this case) but validation accuracy could drop since the smaller dataset is less representative of the overall data so when encountering unseen data it would be more difficult for the model to generalize that data.

2. In this example of spam emails, a false positive would represent an email being predicted as spam when it is not and a false negative would be an email not classified as spam when in reality it is. In terms of which one is worse, a false positive would be worse because an important email could be flagged as spam and thus preventing the email from reaching the user. A false negative would just mean unwanted emails piling up in the users inbox but those emails could easily be manually deleted.

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>DESCRIBE YOUR PROCESS HERE</b></font>

Most of my code was from previous labs. As for areas where I wasn't sure how to
use certain methods or modules etc. I used pandas.pydata.org or w3schools for explanations and documentation, but most of my code could be found in previous lab assignments such as the accuracy scores and splitting the data etc.

I completed the steps in order from top to bottom.

I used generative AI for assistance in finding the missing values. After using it I was still a little confused so I got some help on that part of the question to finally understand.

I struggled a little bit in deciding what would be a good way to fill in missing values but I did some research and quickly found that mean is a valid, ok way to fill them in. Asides from that the rest was okay. Help from the TAs and going through documentation as well as prior labs helped a lot.

# **Part 2: Regression (10.5 marks total)**

| **Question**                               | **Point** |
|--------------------------------------------|-----------|
| **Part 2: Regression**                     |           |
| Step 1: Data Input                         | 1         |
| Step 2: Data Processing                    | 0.5       |
| Step 3: Implement Machine Learning   Model | 1         |
| Step 4: Validate Mode                      | 1         |
| Step 5: Visualize Results                  | 1         |
| Questions                                  | 2         |
| Process Description                        | 4         |
| **Total**                                  | **10.5**  |

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [ ]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets.loaders import load_concrete

X_concrete, y_concrete = load_concrete()

# TO DO: Print size and type of X and y
print(f'Size: {X_concrete.size} Type: {type(X_concrete)}')
print(f'Size: {y_concrete.size} Type: {type(y_concrete)}')

Size: 8240 Type: <class 'pandas.core.frame.DataFrame'>
Size: 1030 Type: <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [ ]:
# TO DO: Check if there are any missing values and fill them in if necessary

# Check for missing values in X_concrete
missing_values_X_concrete = X_concrete.isnull().sum()

# Fill in missing values using mean of each column in X_concrete
X_filled = X_concrete.fillna(X_concrete.mean())

## **Step 3:** Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [ ]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

from sklearn.linear_model import LinearRegression

X_train_concrete, X_test_concrete, y_train_concrete, y_test_concrete = train_test_split(X_concrete, y_concrete, test_size=0.2, random_state=0)

linear_regression_model = LinearRegression();

linear_regression_model.fit(X_train_concrete, y_train_concrete)

LinearRegression()

## **Step 4:** Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [ ]:
# TO DO: ADD YOUR CODE HERE
from sklearn.metrics import r2_score, mean_squared_error

# Making predictions on training set
y_train_concrete_pred = linear_regression_model.predict(X_train_concrete)

# Making predictions on testing set
y_test_concrete_pred = linear_regression_model.predict(X_test_concrete)

# Calculating MSE for training set
mse_train = mean_squared_error(y_train_concrete, y_train_concrete_pred)

# Calculating MSE for testing set
mse_test = mean_squared_error(y_test_concrete, y_test_concrete_pred)

# Calculating R2 Score for training set
r2_train = r2_score(y_train_concrete, y_train_concrete_pred)

# Calculating R2 Score for test set
r2_test = r2_score(y_test_concrete, y_test_concrete_pred)

# Printing results of calculations
print(f"Training Mean Squared Error: {mse_train}")
print(f"Testing Mean Squared Error: {mse_test}")
print(f"Training R² score: {r2_train}")
print(f"Testing R² score: {r2_test}")

Training Mean Squared Error: 110.34550122934108
Testing Mean Squared Error: 95.63533482690434
Training R² score: 0.6090710418548884
Testing R² score: 0.636898110341124


## **Step 5:** Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [ ]:
# TO DO: ADD YOUR CODE HERE
# Prepare the data
data = {
    "Training accuracy": [mse_train, r2_train],
    "Validation accuracy": [mse_test, r2_test]
}

# Create the DataFrame
results = pd.DataFrame(data, index=["MSE", "R2 score"])

results

Training accuracy  Validation accuracy
MSE              110.345501            95.635335
R2 score           0.609071             0.636898

## **Questions (2 marks)**
1. Did using a linear model produce good results for this dataset? Why or why not?

The MSE value was high since a desirable MSE would be one closer to 0. This means the model made significant errors in predictions. In terms of R2 Score the model showed a moderate level of effectiveness in explaining the variance. Thus the model produced a decent fit to the data but could be improved as seen in the high MSE.

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>Explain YOUR PROCESS here:</b></font>

I used the same process to source my code as part 1 except this time since most
of the questions were repeats I just sourced my own code from part 1.

I completed the steps in order from top to bottom.

I used generative AI to help me understand a little more about MSE and R2 score. I was unsure if it was possible to have a mean squared error greater than 100 but it is as seen in this linear model.

I didn't really have challenges in this part because the labs really helped a lot in understanding the assignment and learning important, relevant pandas methods to use.

# **Part 3: Observations/Interpretation (3 marks)**

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


<font color='Green'><b>ADD YOUR FINDINGS HERE</b></font>

The code for both the models were very similar however the main difference is the metrics. Logistic Regression is used for binary outcomes hence the reason it was used in detecting whether an email is spam or not, while linear regression is used for predicting continuous variables. The high MSE indicates the model wasn't overfitting but there is also great room for improvement. The logistic regression model contained


# **Part 4: Reflection (2 marks)**
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


<font color='Green'><b>ADD YOUR THOUGHTS HERE</b></font>

I liked the simplicity of this first assignment since it is our first assignment
we don't have much experience yet with machine learning.

I didn't really dislike anything.

Nothing seemed ambiguous to me in this lab and that is appreciated. It was interesting creating the models using both logistic regression as well as linear regression.

# **Part 5: Bonus Question (4 marks)**

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [ ]:
# TO DO: ADD YOUR CODE HERE

<font color='Green'><b>ADD YOUR ANSWER HERE</b></font>